relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -1,152 +1,228 @@
+    fn get_rule_type(&self) -> RuleType;
+}
 
-        // match against the project
-        if !self.project_ids.is_empty() && !self.project_ids.contains(&project_id) {
-            return false;
+impl FieldValueProvider for Event {
+    fn get_value(&self, field_name: &str) -> FieldValue {
+        match field_name {
+            "event.release" => match self.release.as_str() {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s),
+            },
+            "event.environment" => match self.environment.as_str() {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s),
+            },
+            "event.user_segment" => FieldValue::None, // Not available at this time
+            _ => FieldValue::None,
         }
-
-        // match against the release
-        if !self.releases.is_empty() {
-            match release {
-                None => return false,
-                Some(ref release) => {
-                    if !self.releases.is_match(release) {
-                        return false;
-                    }
-                }
+    }
+    fn get_rule_type(&self) -> RuleType {
+        if let Some(ty) = self.ty.value() {
+            if *ty == EventType::Transaction {
+                return RuleType::Transaction;
             }
         }
+        RuleType::Error
+    }
+}
 
-        // match against the user_segment
-        if !self.user_segments.is_empty() {
-            match user_segment {
-                None => return false,
-                Some(ref user_segment) => {
-                    if !self.user_segments.contains(user_segment) {
-                        return false;
-                    }
-                }
-            }
+impl FieldValueProvider for TraceContext {
+    fn get_value(&self, field_name: &str) -> FieldValue {
+        match field_name {
+            "trace.release" => match &self.release {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s.as_ref()),
+            },
+            "trace.environment" => match &self.environment {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s.as_ref()),
+            },
+            "trace.user_segment" => match &self.user_segment {
+                None => FieldValue::None,
+                Some(s) => FieldValue::String(s.as_ref()),
+            },
+            _ => FieldValue::None,
         }
-        true
+    }
+    fn get_rule_type(&self) -> RuleType {
+        RuleType::Trace
+    }
+}
+
+//TODO make this more efficient
+fn matches<T: FieldValueProvider>(value_provider: &T, rule: &SamplingRule, ty: RuleType) -> bool {
+    if ty != rule.ty {
+        return false;
+    }
+    rule.condition.matches(value_provider)
+}
+
+fn match_glob(rule_val: &GlobPatterns, field_val: &FieldValue) -> bool {
+    match field_val {
+        FieldValue::String(fv) => rule_val.is_match(fv),
```

Comment:
Since you need this match a couple of times, how about introducing a helper method on `FieldValue`:

```rust
impl FieldValue {
  fn as_str(&self) -> Option<&str> {
    // match here
  }
}
```

Then, you can write this, and potentially inline into the call site:

```rust
field_val.as_str().map_or(false, |s| rule_val.is_match(s))
```

Reply:
nice